codex-rs/exec/src/event_processor.rs

```diff
@@ -1,15 +1,35 @@
+use std::path::Path;
+
 use codex_common::summarize_sandbox_policy;
 use codex_core::WireApi;
 use codex_core::config::Config;
 use codex_core::model_supports_reasoning_summaries;
 use codex_core::protocol::Event;
 
+pub(crate) enum CodexStatus {
+    Running,
+    InitiateShutdown,
+    Shutdown,
+}
+
 pub(crate) trait EventProcessor {
     /// Print summary of effective configuration and user prompt.
     fn print_config_summary(&mut self, config: &Config, prompt: &str);
 
     /// Handle a single event emitted by the agent.
-    fn process_event(&mut self, event: Event);
+    fn process_event(&mut self, event: Event) -> CodexStatus;
+
+    /// Get the path to the last message file.
+    fn last_message_path(&self) -> Option<&Path>;
+
+    /// Write the last message to the last message file.
+    fn write_last_message_file(&self, contents: &str) {
+        if let Some(path) = self.last_message_path() {
+            if let Err(e) = std::fs::write(path, contents) {
+                eprintln!("Failed to write last message file {path:?}: {e}");
+            }
+        }
+    }
```

Comment:
Now that both implementations of `EventProcessor` have `last_message_path` as a field, I don't think it makes sense to extend the trait in this way.